src/material/schematics/ng-generate/mdc-migration/rules/components/chips/chips-template.ts

```diff
@@ -0,0 +1,124 @@
+import {replaceStartTag, replaceEndTag, visitElements} from '../../tree-traversal';
+
+export class ChipsTemplateMigrator extends TemplateMigrator {
+  chipInputAttrs: compiler.TmplAstBoundAttribute[] = [];
+
+  getUpdates(node: compiler.TmplAstElement): Update[] {
+    if (node.name === 'mat-chip-list') {
+      return this._buildChipListUpdates(node);
+    }
+
+    if (node.name === 'input') {
+      this._storeChipRefs(node);
+    }
+
+    return [];
+  }
+
+  override onComplete(): void {
+    this.chipInputAttrs = [];
+  }
+
+  /**
+   * Returns start and end tag updates for the given node.
+   *
+   * The updates returned are conditionally determined by whether the given chipListNode is going
+   * to become a mat-chip-grid or a mat-chip-listbox.
+   *
+   * @param node The node to update.
+   * @param chipListNode The node to evaluate as either a grid or listbox.
+   * @param gridTag The new tag name if the chipListNode is a grid.
+   * @param listboxTag The new tag name if the chipListNode is a listbox.
+   * @returns The start and end tag updates for the given node.
+   */
+  private _buildUpdates(
+    node: compiler.TmplAstElement,
+    chipListNode: compiler.TmplAstElement,
+    gridTag: string,
+    listboxTag: string,
+  ): Update[] {
+    return [
+      {
+        location: node.startSourceSpan.start,
+        updateFn: html => {
+          return this._isChipGrid(chipListNode)
+            ? replaceStartTag(html, node, gridTag)
+            : replaceStartTag(html, node, listboxTag);
+        },
+      },
+      {
+        location: node.endSourceSpan!.start,
+        updateFn: html => {
+          return this._isChipGrid(chipListNode)
+            ? replaceEndTag(html, node, gridTag)
+            : replaceEndTag(html, node, listboxTag);
+        },
+      },
+    ];
+  }
+
+  /** Stores the matChipInputFor references on given input. */
+  private _storeChipRefs(node: compiler.TmplAstElement): void {
+    for (let i = 0; i < node.inputs.length; i++) {
+      if (node.inputs[i].name === 'matChipInputFor') {
+        this.chipInputAttrs.push(node.inputs[i]);
+      }
+    }
+  }
+
+  /**
+   * Builds and returns the updates for the given
+   * mat-chip-list node as well as any mat-chip-lists it contains.
+   */
+  private _buildChipListUpdates(node: compiler.TmplAstElement): Update[] {
+    return [
+      ...this._buildUpdates(node, node, 'mat-chip-grid', 'mat-chip-listbox'),
+      ...this._buildChipUpdates(node),
+    ];
+  }
+
+  /** Builds and returns the updates for the mat-chips inside the given mat-chip-list node. */
+  private _buildChipUpdates(node: compiler.TmplAstElement): Update[] {
+    const updates: Update[] = [];
+
+    // Recursively check the children of the mat-chip-list for mat-chip elements.
+    const handleMatChipUpdates = (child: compiler.TmplAstElement) => {
+      if (child.name !== 'mat-chip') {
+        visitElements(child.children, handleMatChipUpdates);
```

Comment:
The fact that we recursively would visit elements here, along with `getUpdates` (which is non-ideal -- perf wise likely not an issue) seems to indicate that it might be better just letting the migrators traverse a template on their own? Also that we would need to have some kind of resetting mechanism seems avoidable like that.

e.g.

```ts
abstract getUpdatesForTemplate(tmpl: AST) {
  // with a convenience function like:
  return this.visitEveryElementAndCollectUpdates(tmpl, 
    el => this._getListUpdatesForElement(el))
}
```

Not feeling strongly about this though, just an idea

Reply:
I was bothered by this, too, so I did another kinda big refactor to get things working in only a single pass. I had to expose both the preorder and postorder callback to get things working and changed up the coding paradigm so that we return all of the updates after the tree traversal, not during it. Let me know if things are getting too crazy lol